### PR TITLE
split Balances Get/Put interface methods into separate methods for creatables

### PR DESCRIPTION
**Pull request by cce**
## Summary

This adds extra methods to the apply.Balances interface to offer creatable-specific alternatives to Get and Put, including chasing down all the use cases of Get and Put to call the new interface methods. A new apply.AccountData type restricts access to the full basics.AccountData structure.

In later PRs, the goal is to replace basics.AccountData and migrate to a database that uses separate tables for account data and per-(account, app/asset) data. 

## Test Plan

All the current tests pass. I'm also running it against mainnet and testnet with CatchpointTracking enabled and comparing the resulting catchpoint hashes with historical catchpoint hash values. I'm also looking at coverage of files in apply to see if there are any new tests I could add to master (and then merge into this branch) to cover any corner cases not currently handled by our existing unit tests.

### CODE REVIEW

**Review by algorandskiy — COMMENTED** [review]
Looks good. Please fix the build, comment new functions, and it is ready to go

**Comment by cce** [comment]
To try and catch anything I missed, I prefixed the maps in AccountData with `X` and chased down all the users in the ledger package: https://github.com/cce/go-algorand/commit/1b917836952e6f6a935389d7d9709887f12ccee7#diff-831e406846493c27b22aca1b7482743df94cec8a6957b34059c912ff7996ec28R464

I found [`AccountData.MinBalance()`](https://github.com/cce/go-algorand/commit/1b917836952e6f6a935389d7d9709887f12ccee7#diff-17a1b4b06e73ea379f82020cc8574075f8cef91af858a37f0524a6e713329b1cR94) which counts the size of Assets, AppParams, and AppLocalStates, which is called by logicLedger.MinBalance() and [BlockEvaluator.checkMinBalance()](https://github.com/cce/go-algorand/commit/1b917836952e6f6a935389d7d9709887f12ccee7#diff-9f52ec3f95afe73702fa021a8d62dfeb6e192b3cf4883aa10863ad20d815fe3cR837). I [modified AccountData.MinBalance](https://github.com/cce/go-algorand/commit/1b917836952e6f6a935389d7d9709887f12ccee7#diff-831e406846493c27b22aca1b7482743df94cec8a6957b34059c912ff7996ec28R464) to accept an external totalAssets, totalAppParams, totalAppLocalStates rather than looking it up from maps.

WDYT — I was thinking of moving AccountData.MinBalance() to an internal helper function in ledger called by both callers, would that be a good idea?

File: ledger/apply/asset.go [hunk 1]
```diff
@@ -259,13 +252,16 @@ func AssetTransfer(ct transactions.AssetTransferTxnFields, header transactions.H
 
 	// Allocate a slot for asset (self-transfer of zero amount).
 	if ct.AssetAmount == 0 && ct.AssetReceiver == source && !clawback {
-		snd, err := balances.Get(source, false)
-		if err != nil {
+		sndHolding, err := balances.GetAssetHolding(source, ct.XferAsset)
```
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/asset.go:255]: Hrmm, yes, that seems better — will make the ok-checking logic easier to read and make it harder to miss checking errors.Is(ErrCreatableNotFound). Will fix

File: ledger/apply/application.go [hunk 2]
```diff
@@ -221,20 +228,27 @@ func optInApplication(balances Balances, sender basics.Address, appIdx basics.Ap
 	}
 
 	// If the user has already opted in, fail
-	_, ok := record.AppLocalStates[appIdx]
-	if ok {
+	ok, err := balances.CheckAppLocalState(sender, appIdx)
+	if err != nil {
+		return err
+	}
+	if !ok {
```
Comment (cce) [inline ledger/apply/application.go:235]: caught this bug

File: ledger/apply/asset.go [hunk 3]
```diff
@@ -406,18 +396,10 @@ func AssetFreeze(cf transactions.AssetFreezeTxnFields, header transactions.Heade
 	}
 
 	// Get the account to be frozen/unfrozen.
-	record, err := balances.Get(cf.FreezeAccount, false)
+	holding, err := balances.GetAssetHolding(cf.FreezeAccount, cf.FreezeAsset)
```
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/asset.go]: how so?

[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/apply/asset.go]: I'd expect `GetAssetHolding` to return holdings by address (`cf.FreezeAccount` in this case), but looks like it also checks/sets `cf.FreezeAsset` value

File: ledger/internal/cow_creatables.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+	return len(acct.AppParams), nil
+}
+func (cs *roundCowState) TotalAppLocalState(addr basics.Address) (int, error) {
+	acct, err := cs.lookup(addr)
+	if err != nil {
+		return 0, err
+	}
+	return len(acct.AppLocalStates), nil
+}
+func (cs *roundCowState) TotalAssetHolding(addr basics.Address) (int, error) {
```
Comment (jannotti) [inline ledger/internal/cow_creatables.go:46]: Since I previously said I'd prefer AssetHoldingCount, I suppose I should say I agree they should all be uniform.  AppParamsCount, AppLocalStateCount seem ok.

File: ledger/internal/cow_creatables.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	m := make(map[basics.AssetIndex]basics.AssetParams, len(acct.AssetParams))
+	for k, v := range acct.AssetParams {
+		m[k] = v
+	}
+	delete(m, aidx)
+	acct.AssetParams = m
+	return cs.putAccount(addr, acct)
+}
+
+func (cs *roundCowState) CheckAppLocalState(addr basics.Address, aidx basics.AppIndex) (ok bool, err error) {
```
Comment (jannotti) [inline ledger/internal/cow_creatables.go:200]: Does this exist because we're planning ahead for a db layer that makes checking existence faster than returning the value?  Or to avoid the function call overhead of returning the value?

**Review by jannotti — COMMENTED** [review]
Had a few questions, but looks like a nice step forward to me.

File: ledger/apply/application.go [hunk 6]
```diff
@@ -86,19 +85,22 @@ func createApplication(ac *transactions.ApplicationCallTxnFields, balances Balan
 		return
 	}
 
+	// look up how many apps they have
+	totalAppParams, err := balances.TotalAppParams(creator)
```
[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/apply/application.go:89]: I now see that, at least for now, you're not caching a total count.  I wonder if we should.  The reasoning would be similar, but it does mean writing these objects (man, I really want them to have another name like "AccountHeader" or "AccountBase" so I don't have to distinguish by package name), whenever we add or remove an asset or app.  Then again, I guess we already have to, exactly so we can calc min balance.

File: ledger/apply/apply.go [hunk 7]
```diff
@@ -18,21 +18,102 @@ package apply
...
+	VoteKeyDilution uint64
+
+	AuthAddr           basics.Address
+	TotalAppSchema     basics.StateSchema
+	TotalExtraAppPages uint32
+}
+
+// ToApplyAccountData returns apply.AccountData from basics.AccountData
+func ToApplyAccountData(acct basics.AccountData) AccountData {
+	return AccountData{
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/apply/apply.go:48]: I guess `apply.AccountData` is a temporary type while we do not have other/better types.
The idea is to have in-mem and persistent data types for components of `basics.AccountData`

File: ledger/internal/cow_creatables.go (hunk 4 above)
Comment (algorandskiy) [reply ledger/internal/cow_creatables.go:46]: This `Total*` name comes from existing `TotalAppSchema/TotalExtraAppPages` and upcoming `TotalAssets/TotalAppParams/TotalAppLocalStates` fields

File: ledger/internal/cow_creatables.go (hunk 5 above)
Comment (algorandskiy) [reply ledger/internal/cow_creatables.go:200]: Empty local state is can exist (empty schema + empty KV) and this is a valid, so the `ok` flag would be needed anyway. Plus, existence check does not perform deserialization so it is a bit faster.

Eventually it will be a two steps check: 1) modified data, 2) a base cow/ledger caches and `COUNT` query.

File: ledger/apply/apply.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/apply/apply.go:48]: I get that, but it seems to me that this in-memory type is, in fact, the beginning of basics.AccountData, so why not make that explicit?

File: ledger/internal/cow_creatables.go (hunk 4 above)
Comment (jannotti) [reply ledger/internal/cow_creatables.go:46]: But notice that TotalAppSchema and TotalExtraPages are actually _totaling_.  They add up the values inside the schema, or the values of the epp fields.  These don't total up anything. But, since one is inclined to think that a Total*() method is totaling things up, when you see TotalAssetHolding it seems like it must be totaling up some assets.

File: ledger/internal/cow_creatables.go (hunk 5 above)
Comment (cce) [reply ledger/internal/cow_creatables.go:200]: Besides SQL count and copying values, there are also potential performance benefits in different KV stores that would make checking existence faster than fetching & returning a copy of the value, but the basic thing I was trying to capture with the extra CheckX methods was what's the absolute minimum needed from the account DB when evaluating. Maybe HasX is a better name though..

File: ledger/apply/apply.go [hunk 8]
```diff
@@ -18,21 +18,102 @@ package apply
...
+	PutAppParams(addr basics.Address, aidx basics.AppIndex, params basics.AppParams) error
+	DeleteAppParams(addr basics.Address, aidx basics.AppIndex) error
+
+	TotalAppLocalState(addr basics.Address) (int, error)
+	GetAppLocalState(addr basics.Address, aidx basics.AppIndex) (basics.AppLocalState, bool, error)
+	CheckAppLocalState(addr basics.Address, aidx basics.AppIndex) (bool, error)
+	PutAppLocalState(addr basics.Address, aidx basics.AppIndex, state basics.AppLocalState) error
+	DeleteAppLocalState(addr basics.Address, aidx basics.AppIndex) error
+
+	TotalAssetHolding(addr basics.Address) (int, error)
```
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go:107]: CountAssetHolding?

File: ledger/internal/cow_creatables.go (hunk 5 above)
Comment (jannotti) [reply ledger/internal/cow_creatables.go:200]: I'm totally on board with the purpose, I was mostly checking to make sure that's why it existed.

I do like either HasXXX() or XXXExists() though, as CheckXXX() sounds like it's doing a sanity check or logic check on the values.

File: ledger/apply/apply.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go:48]: There was a bigger potential change I considered, where basics.AccountData loses these maps altogether, ledger.Lookup() gets split into separate acct/app/asset methods too, all former users of AccountData are updated to get asset data separately, and the original AccountData type (and its msgp encoding) is moved inside ledger and becomes an implementation detail of account DB, eval, COW, etc. 

I didn't think of embedding it as an AccountBase in AccountData because I figured we this PR was moving us towards that goal, starting with apply.Balances.

However looking at what's missing, really only the REST API is left after this I believe.

[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go:48]: oh, also I was worried a change to basics.AccountData directly might involving updating a ton of tests.

File: ledger/apply/asset.go [hunk 9]
```diff
@@ -70,32 +69,37 @@ func getParams(balances Balances, aidx basics.AssetIndex) (params basics.AssetPa
...
-		record.AssetParams[newidx] = cc.AssetParams
-		record.Assets[newidx] = basics.AssetHolding{
+		assetParams := cc.AssetParams
+		assetHolding := basics.AssetHolding{
 			Amount: cc.AssetParams.Total,
 		}
 
-		if len(record.Assets) > balances.ConsensusParams().MaxAssetsPerAccount {
-			return fmt.Errorf("too many assets in account: %d > %d", len(record.Assets), balances.ConsensusParams().MaxAssetsPerAccount)
+		if totalAssets >= balances.ConsensusParams().MaxAssetsPerAccount {
```
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/asset.go:94]: yeah, I was overoptimizing for min diff length, I think there are a few places like this I can pull the "too many X" checks up

File: ledger/apply/apply.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/apply/apply.go:48]: I see, so the plan might be that basics.AccountData becomes what I was calling basics.AccountBase, and it gets (maybe) embedded in ledger.accountFull.  The apply package will take the various pieces it needs, never an entire ledger.accountFull.

File: ledger/apply/apply.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/apply/apply.go:107]: Fine by me. I read that as verb & noun where putting it at the end feels more like a noun phrase, which seems closer to reality (there's no actual counting going on). But that's a nitpick that I don't think matters as much as conveying "Count".

[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go:107]: with "unlimited" assets/apps, you will end up with a SQL count too. also Count/Has/Get/Put/Delete is the new CRUD

File: ledger/apply/apply.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go:48]: y'know, I already changed cowForLogicLedger to have Get() return the new apply.AccountData, which is now out of sync with Ledger.Get() returning basics.AccountData, so I will try plumbing through Ledger.GetAppParams, etc to see how much more changes it would be.

[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/apply/apply.go:48]: Ok, don't get carried away on my behalf.  This is nice enough to do as is, and I don't want it to get so large we we have trouble moving forward incrementally.

File: ledger/apply/asset.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/apply/asset.go:94]: No strong need, sometimes I don't really notice what's new and what was already there.

File: ledger/apply/apply.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go:48]: I tried tracking down everywhere that was using the maps in basics.AccountData, it was the REST API (account data and dryrun endpoints), tealdbg, and goal, all calling through Ledger.Lookup() basically. I think it's probably best for now to have ledger.Ledger just hide the logic behind Lookup() of joining the asset/app data with account data, rather than change all the tests and callers to Lookup that use basics.AccountData.

[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go:48]: That's in: https://github.com/cce/go-algorand/compare/split-balances...split-balances-hide-accountdata-maps

File: ledger/apply/apply.go [hunk 10]
```diff
@@ -18,21 +18,102 @@ package apply
...
+
+	a.AuthAddr = acct.AuthAddr
+	a.TotalAppSchema = acct.TotalAppSchema
+	a.TotalExtraAppPages = acct.TotalExtraAppPages
+}
+
 // Balances allow to move MicroAlgos from one address to another and to update balance records, or to access and modify individual balance records
 // After a call to Put (or Move), future calls to Get or Move will reflect the updated balance record(s)
 type Balances interface {
 	// Get looks up the account data for an address, ignoring application state
```
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go:87]: That's funny, this comment was added maybe by @algorandskiy in https://github.com/algorand/go-algorand/pull/1765/files#diff-18c6180c8896bfd593a004408be3d01ce69f87ea36a7b9b379665d155acf3cd3R28 and then Balances got updated again to remove PutWithCreatable in https://github.com/algorand/go-algorand/pull/2638/files#diff-18c6180c8896bfd593a004408be3d01ce69f87ea36a7b9b379665d155acf3cd3L36 so I suspect this comment is out of date and reflected an older version of Get()?

File: ledger/apply/apply.go [hunk 11]
```diff
@@ -18,21 +18,102 @@ package apply
...
 import (
 	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
 	"github.com/algorand/go-algorand/data/basics"
 	"github.com/algorand/go-algorand/data/transactions"
 	"github.com/algorand/go-algorand/data/transactions/logic"
 )
 
+// AccountData is like basics.AccountData, but without any maps
+// containing AppParams, AppLocalState, AssetHolding, or AssetParams.
```
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go]: Updated with more expository content

File: ledger/apply/apply.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (cce) [reply ledger/apply/apply.go]: @algorandskiy I updated the comment.

**Review by algorandskiy — APPROVED** [review]
